---
title: "DRAGONsc"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using DRAGONsc for scRNAseq clustering}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---
# DRAGONsc for single-cell RNAseq clustering

DeteRminisitic Annealing Gaussian mixture mOdels for clusteriNg Single-Cell RNAseq (DRAGONsc) data is a clustering algorithm for single-cell RNAseq built upon the statistical framework of Gaussian mixture models and the concept of deterministic annealing. Analagous to the physical process of annealing, DRAGON uses a temperature parameter to put a constraint on the minimum entropy of the clustering solution. At each temperature, DRAGON performs maximum likelihood estimation to ascertain the paramaters of a Gaussian mixture model at that temperature. This allows us to define the "strongest" divisions in the data at higher temperatures, and to reveal more subtle differences as we gradually cool the temperature. This framework also allows us to split clusters in a statistically meaningful way as we reduce the temperature.

If you're into the math behind this, check out our publication (coming to a journal near you soon), and the great work of Ken Rose.

Here, we demonstrate standard use of DRAGON on a simulated dataset. We do this in the following steps:

- Simulate data using the R package *splatter*
- Process our data using a standard Seurat workflow (v.2.3.4)
- Use DRAGON to cluster our simulated data
- Compare the clustering results from DRAGON to the ground truth

# Simulate data using splatter

First, we simulate data to use with splatter. Alternatively we can bypass this step with a pre-simulated dataset that I've included with the package.

```{r}
suppressMessages({
library(splatter)
library(dragonsc)
library(Seurat)
})

#Create parameters and simulate data
sc.params <- newSplatParams(nGenes=1000,batchCells=5000)
sim.groups <- splatSimulate(params=sc.params,method="groups",group.prob=c(0.10,0.10,0.15,0.15,0.25,0.25),de.prob=c(0.3,0.2,0.2,0.1,0.2,0.1),verbose=F)
sim.groups
```

This will create some simulated data consisting of 6 clusters, with various fractions of genes differentially expressed between clusters.

# Seurat workflow on simulated data

Now that we've simlulated some data, we will pass it through a standard Seurat workflow (using v2.3.4). Our goal is to generate a principal component analysis, and determine how many principal compnents we will use include for DRAGON cluster. We will also create a TSNE plot to visualize the clusters later.

```{r,fig.width=7,fig.height=7,fig.align="center"}
#Pass simulated data to Seurat
ser <- CreateSeuratObject(raw.data=assays(sim.groups)$counts)

#Normal Seurat workflow
ser <- NormalizeData(ser)
ser <- ScaleData(ser)
ser@var.genes <- rownames(ser@raw.data)
#We will pretend like the 1000 simulated genes are the "variable genes",
#and we will skip FindVariableGenes from Seurat
ser <- RunPCA(ser,do.print=FALSE)

PCElbowPlot(ser)
PCHeatmap(ser,pc.use=1:10,cells.use=500,do.balanced=T)

#Looks like the first 6 PCs capture most of the variance
#We will include the first 7 since we like to live dangerously

ser <- RunTSNE(ser,dims.use=1:7)

#Add the simulated data IDs to the Seurat object
data.to.add <- colData(sim.groups)$Group
names(data.to.add) <- ser@cell.names
ser <- AddMetaData(ser,metadata=data.to.add,col.name="real_id")
```

# DRAGON clustering

With the PCA done, we will now perfrom the DRAGON clustering using the reduced dimensionality of the data. But first, we need to determine a starting temperature.

```{r,fig.width=7,fig.height=3.5,fig.align="center"}
#Seurat stores PCA results in the dimensionality reduction slot "dr".
pcs.for.dragon <- ser@dr$pca@cell.embeddings[,1:7]

#Our criterion for splitting a cluster is 2 times the first
#eigenvalue of the covariance matrix, as per below
est.temp <- eigen(cov(pcs.for.dragon))$values[[1]]*2

#We will strat 10% below the value of est.temp
int.temp <- est.temp-(est.temp*0.1)

#Next, we set up a series of annealing steps where we progressively decrease the temperature.
annealing.steps <- data.frame("step"=cbind(1+c(0:11)),"temperature"=c(int.temp*exp(-(0:10)*0.45),0))

#Then we run the algorithm! It will create a folder called "dragon_intermediate_files" since save.intermediate.files=TRUE

setwd("/Users/arc85/Desktop/")

overall.clustering.soln <- dragonsc(pca.components=pcs.for.dragon,temp.decay.steps=annealing.steps,max.iterations=20,delta.log.likelihood=1e-6,max.clusters=6,save.intermediate.files=TRUE,num.cores=2,verbose=TRUE)

#Add our results back into the Seurat object
drag.clust <- overall.clustering.soln[["clusters"]]
data.to.add <- vector(length=length(ser@cell.names))
names(data.to.add) <- ser@cell.names
data.to.add <- overall.clustering.soln[["clusters"]]
ser <- AddMetaData(ser,metadata=data.to.add,col.name="DRAGON")

#Visualize the final result vs the ground truth
p1 <- TSNEPlot(ser,group.by="real_id",do.return=T) + ggtitle("Ground truth")
p2 <- TSNEPlot(ser,group.by="DRAGON",do.return=T) + ggtitle("DRAGON")
plot_grid(p1,p2)
```

Congratulations! You have performed your first clustering with DRAGON. Note that the clusters don't align by name. Let's fix that with BuildClusterTree in Seurat.

```{r,fig.width=7,fig.height=3.5,fig.align="center"}
ser <- SetAllIdent(ser,id="real_id")
ser <- BuildClusterTree(ser,do.reorder=T,reorder.numeric=T,do.plot=FALSE,show.progress=FALSE)
ser@meta.data$real_id <- ser@meta.data$tree.ident

ser <- SetAllIdent(ser,id="DRAGON")
ser <- BuildClusterTree(ser,do.reorder=T,reorder.numeric=T,do.plot=FALSE,show.progress=FALSE)
ser@meta.data$DRAGON <- ser@meta.data$tree.ident

p1 <- TSNEPlot(ser,group.by="real_id",do.return=T) + ggtitle("Ground truth")
p2 <- TSNEPlot(ser,group.by="DRAGON",do.return=T) + ggtitle("DRAGON")
plot_grid(p1,p2)
```

# Compare results using confusion matrices from caret

We can quantify the similarity between Ground truth and DRAGON using a confusion matrix from the caret package.

```{r}
suppressMessages(library(caret))
caret::confusionMatrix(data=as.factor(ser@meta.data$DRAGON),reference=as.factor(ser@meta.data$real_id))
```

# View progressive clustering results as the temperature is decreased

If you've chosen to save the intermediate results (as we have here by setting save.intermediate.results=TRUE), you can check out how clusters form as the temperature of the system is reduced. Annealing is cool! (Pun intended)

```{r,fig.width=7,fig.height=7,fig.align="center"}
setwd("/Users/arc85/Desktop/dragon_intermediate_files")

temp.60 <- readRDS("em_results_temp_60.RDS")
names(temp.60)
#Clustering results are present in the "clusters" item in the named list

drag.clust <- temp.60[["clusters"]]
names(drag.clust) <- ser@cell.names
ser <- AddMetaData(ser,metadata=drag.clust,col.name="temp_60")

temp.4 <- readRDS("em_results_temp_4.RDS")
drag.clust <- temp.4[["clusters"]]
names(drag.clust) <- ser@cell.names
ser <- AddMetaData(ser,metadata=drag.clust,col.name="temp_4")

temp.1 <- readRDS("em_results_temp_1.RDS")
drag.clust <- temp.1[["clusters"]]
names(drag.clust) <- ser@cell.names
ser <- AddMetaData(ser,metadata=drag.clust,col.name="temp_1")

p3 <- TSNEPlot(ser,group.by="temp_60",do.return=T) + ggtitle("Temperature 60")
p4 <- TSNEPlot(ser,group.by="temp_4",do.return=T) + ggtitle("Temperature 4")
p5 <- TSNEPlot(ser,group.by="temp_1",do.return=T) + ggtitle("Temperature 1")
p6 <- TSNEPlot(ser,group.by="DRAGON",do.return=T) +
ggtitle("Temperature 0")

plot_grid(p3,p4,p5,p6,ncol=2)
```
