<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>DRAGONsc • dragonsc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="DRAGONsc">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dragonsc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/dragon.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>DRAGONsc</h1>
            
      
      
      <div class="hidden name"><code>dragonsc.Rmd</code></div>

    </div>

    
    
<div id="dragonsc-for-single-cell-rnaseq-clustering" class="section level1">
<h1 class="hasAnchor">
<a href="#dragonsc-for-single-cell-rnaseq-clustering" class="anchor"></a>DRAGONsc for single-cell RNAseq clustering</h1>
<p>DeteRminisitic Annealing Gaussian mixture mOdels for clusteriNg Single-Cell RNAseq (DRAGONsc) data is a clustering algorithm for single-cell RNAseq built upon the statistical framework of Gaussian mixture models and the concept of deterministic annealing. Analagous to the physical process of annealing, DRAGON uses a temperature parameter to put a constraint on the minimum entropy of the clustering solution. At each temperature, DRAGON performs maximum likelihood estimation to ascertain the paramaters of a Gaussian mixture model at that temperature. This allows us to define the “strongest” divisions in the data at higher temperatures, and to reveal more subtle differences as we gradually cool the temperature. This framework also allows us to split clusters in a statistically meaningful way as we reduce the temperature.</p>
<p>If you’re into the math behind this, check out our publication (coming to a journal near you soon), and the great work of Ken Rose.</p>
<p>Here, we demonstrate standard use of DRAGON on a simulated dataset. We do this in the following steps:</p>
<ul>
<li>Simulate data using the R package <em>splatter</em>
</li>
<li>Process our data using a standard Seurat workflow (v.2.3.4)</li>
<li>Use DRAGON to cluster our simulated data</li>
<li>Compare the clustering results from DRAGON to the ground truth</li>
</ul>
</div>
<div id="simulate-data-using-splatter" class="section level1">
<h1 class="hasAnchor">
<a href="#simulate-data-using-splatter" class="anchor"></a>Simulate data using splatter</h1>
<p>First, we simulate data to use with splatter. Alternatively we can bypass this step with a pre-simulated dataset that I’ve included with the package.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressMessages</a></span>({</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(splatter)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dragonsc)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(Seurat)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5">})</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">#Create parameters and simulate data</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8">sc.params &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/splatter/topics/newParams">newSplatParams</a></span>(<span class="dt">nGenes=</span><span class="dv">1000</span>,<span class="dt">batchCells=</span><span class="dv">5000</span>)</a>
<a class="sourceLine" id="cb1-9" data-line-number="9">sim.groups &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/splatter/topics/splatSimulate">splatSimulate</a></span>(<span class="dt">params=</span>sc.params,<span class="dt">method=</span><span class="st">"groups"</span>,<span class="dt">group.prob=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.10</span>,<span class="fl">0.10</span>,<span class="fl">0.15</span>,<span class="fl">0.15</span>,<span class="fl">0.25</span>,<span class="fl">0.25</span>),<span class="dt">de.prob=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.3</span>,<span class="fl">0.2</span>,<span class="fl">0.2</span>,<span class="fl">0.1</span>,<span class="fl">0.2</span>,<span class="fl">0.1</span>),<span class="dt">verbose=</span>F)</a>
<a class="sourceLine" id="cb1-10" data-line-number="10">sim.groups</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 1000 5000 
## metadata(1): Params
## assays(6): BatchCellMeans BaseCellMeans ... TrueCounts counts
## rownames(1000): Gene1 Gene2 ... Gene999 Gene1000
## rowData names(10): Gene BaseGeneMean ... DEFacGroup5 DEFacGroup6
## colnames(5000): Cell1 Cell2 ... Cell4999 Cell5000
## colData names(4): Cell Batch Group ExpLibSize
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<p>This will create some simulated data consisting of 6 clusters, with various fractions of genes differentially expressed between clusters.</p>
</div>
<div id="seurat-workflow-on-simulated-data" class="section level1">
<h1 class="hasAnchor">
<a href="#seurat-workflow-on-simulated-data" class="anchor"></a>Seurat workflow on simulated data</h1>
<p>Now that we’ve simlulated some data, we will pass it through a standard Seurat workflow (using v2.3.4). Our goal is to generate a principal component analysis, and determine how many principal compnents we will use include for DRAGON cluster. We will also create a TSNE plot to visualize the clusters later.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">#Pass simulated data to Seurat</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/CreateSeuratObject">CreateSeuratObject</a></span>(<span class="dt">raw.data=</span><span class="kw">assays</span>(sim.groups)<span class="op">$</span>counts)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">#Normal Seurat workflow</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/NormalizeData">NormalizeData</a></span>(ser)</a>
<a class="sourceLine" id="cb3-6" data-line-number="6">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/ScaleData">ScaleData</a></span>(ser)</a></code></pre></div>
<pre><code>## Scaling data matrix</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">ser<span class="op">@</span>var.genes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(ser<span class="op">@</span>raw.data)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co">#We will pretend like the 1000 simulated genes are the "variable genes",</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="co">#and we will skip FindVariableGenes from Seurat</span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/RunPCA">RunPCA</a></span>(ser,<span class="dt">do.print=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-5" data-line-number="5"></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/PCElbowPlot">PCElbowPlot</a></span>(ser)</a></code></pre></div>
<p><img src="dragonsc_files/figure-html/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/PCHeatmap">PCHeatmap</a></span>(ser,<span class="dt">pc.use=</span><span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,<span class="dt">cells.use=</span><span class="dv">500</span>,<span class="dt">do.balanced=</span>T)</a></code></pre></div>
<p><img src="dragonsc_files/figure-html/unnamed-chunk-2-2.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co">#Looks like the first 6 PCs capture most of the variance</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="co">#We will include the first 7 since we like to live dangerously</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"></a>
<a class="sourceLine" id="cb7-4" data-line-number="4">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/RunTSNE">RunTSNE</a></span>(ser,<span class="dt">dims.use=</span><span class="dv">1</span><span class="op">:</span><span class="dv">7</span>)</a>
<a class="sourceLine" id="cb7-5" data-line-number="5"></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="co">#Add the simulated data IDs to the Seurat object</span></a>
<a class="sourceLine" id="cb7-7" data-line-number="7">data.to.add &lt;-<span class="st"> </span><span class="kw">colData</span>(sim.groups)<span class="op">$</span>Group</a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(data.to.add) &lt;-<span class="st"> </span>ser<span class="op">@</span>cell.names</a>
<a class="sourceLine" id="cb7-9" data-line-number="9">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/AddMetaData">AddMetaData</a></span>(ser,<span class="dt">metadata=</span>data.to.add,<span class="dt">col.name=</span><span class="st">"real_id"</span>)</a></code></pre></div>
</div>
<div id="dragon-clustering" class="section level1">
<h1 class="hasAnchor">
<a href="#dragon-clustering" class="anchor"></a>DRAGON clustering</h1>
<p>With the PCA done, we will now perfrom the DRAGON clustering using the reduced dimensionality of the data. But first, we need to determine a starting temperature.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co">#Seurat stores PCA results in the dimensionality reduction slot "dr".</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2">pcs.for.dragon &lt;-<span class="st"> </span>ser<span class="op">@</span>dr<span class="op">$</span>pca<span class="op">@</span>cell.embeddings[,<span class="dv">1</span><span class="op">:</span><span class="dv">7</span>]</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co">#Our criterion for splitting a cluster is 2 times the first</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="co">#eigenvalue of the covariance matrix, as per below</span></a>
<a class="sourceLine" id="cb8-6" data-line-number="6">est.temp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/eigen">eigen</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/cor">cov</a></span>(pcs.for.dragon))<span class="op">$</span>values[[<span class="dv">1</span>]]<span class="op">*</span><span class="dv">2</span></a>
<a class="sourceLine" id="cb8-7" data-line-number="7"></a>
<a class="sourceLine" id="cb8-8" data-line-number="8"><span class="co">#We will strat 10% below the value of est.temp</span></a>
<a class="sourceLine" id="cb8-9" data-line-number="9">int.temp &lt;-<span class="st"> </span>est.temp<span class="op">-</span>(est.temp<span class="op">*</span><span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb8-10" data-line-number="10"></a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="co">#Next, we set up a series of annealing steps where we progressively decrease the temperature.</span></a>
<a class="sourceLine" id="cb8-12" data-line-number="12">annealing.steps &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="st">"step"</span>=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(<span class="dv">1</span><span class="op">+</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span><span class="op">:</span><span class="dv">11</span>)),<span class="st">"temperature"</span>=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(int.temp<span class="op">*</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">exp</a></span>(<span class="op">-</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">10</span>)<span class="op">*</span><span class="fl">0.45</span>),<span class="dv">0</span>))</a>
<a class="sourceLine" id="cb8-13" data-line-number="13"></a>
<a class="sourceLine" id="cb8-14" data-line-number="14"><span class="co">#Then we run the algorithm! It will create a folder called "dragon_intermediate_files" since save.intermediate.files=TRUE</span></a>
<a class="sourceLine" id="cb8-15" data-line-number="15"></a>
<a class="sourceLine" id="cb8-16" data-line-number="16"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"/Users/arc85/Desktop/"</span>)</a>
<a class="sourceLine" id="cb8-17" data-line-number="17"></a>
<a class="sourceLine" id="cb8-18" data-line-number="18">overall.clustering.soln &lt;-<span class="st"> </span><span class="kw">dragonsc</span>(<span class="dt">pca.components=</span>pcs.for.dragon,<span class="dt">temp.decay.steps=</span>annealing.steps,<span class="dt">max.iterations=</span><span class="dv">20</span>,<span class="dt">delta.log.likelihood=</span><span class="fl">1e-6</span>,<span class="dt">max.clusters=</span><span class="dv">6</span>,<span class="dt">save.intermediate.files=</span><span class="ot">TRUE</span>,<span class="dt">num.cores=</span><span class="dv">2</span>,<span class="dt">verbose=</span><span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] "Initializing DRAGON clustering at temperature=60"
## [1] "Starting loop 1 at temperature=60"
## [1] "Current log likelihood=-5836.81597342153"
## [1] "Current log likelihood=-5836.42383327996"
## [1] "Current log likelihood=-5836.32482152013"
## [1] "Current log likelihood=-5836.2940279079"
## [1] "Current log likelihood=-5836.2838051644"
## [1] "Current log likelihood=-5836.28005327516"
## [1] "Current log likelihood=-5836.27834691306"
## [1] "Current log likelihood=-5836.27723866244"
## [1] "Current log likelihood=-5836.27622655925"
## [1] "Current log likelihood=-5836.27511209313"
## [1] "Current log likelihood=-5836.27379173256"
## [1] "Current log likelihood=-5836.27218684643"
## [1] "Current log likelihood=-5836.27021830949"
## [1] "Current log likelihood=-5836.26779546624"
## [1] "Current log likelihood=-5836.26480943315"
## [1] "Current log likelihood=-5836.26112728852"
## [1] "Current log likelihood=-5836.25658583258"
## [1] "Current log likelihood=-5836.25098428177"
## [1] "Current log likelihood=-5836.24407545874"
## [1] "Current log likelihood=-5836.2355550679"
## [1] "Completed loop 1 at temperature=60"
## [1] "Starting loop 2 at temperature=4"
## [1] "Current log likelihood=-97679.6442332053"
## [1] "Current log likelihood=-90171.5695400773"
## [1] "Current log likelihood=-78115.0538687444"
## [1] "Current log likelihood=-61278.9484706004"
## [1] "Current log likelihood=-61219.1516504178"
## [1] "Current log likelihood=-61219.1516504167"
## [1] "Completed loop 2 at temperature=4"
## [1] "Starting loop 3 at temperature=2.6"
## [1] "Current log likelihood=-93635.0706079769"
## [1] "Current log likelihood=-46182.319236626"
## [1] "Current log likelihood=-14013.6472392844"
## [1] "Current log likelihood=-14013.6472320344"
## [1] "Current log likelihood=-14013.6472320344"
## [1] "Completed loop 3 at temperature=2.6"
## [1] "Starting loop 4 at temperature=1.6"
## [1] "Current log likelihood=-21977.7737171601"
## [1] "Current log likelihood=-21977.7737171601"
## [1] "Completed loop 4 at temperature=1.6"
## [1] "Starting loop 5 at temperature=1"
## [1] "Current log likelihood=-34468.0103305677"
## [1] "Current log likelihood=-34468.0103305677"
## [1] "Completed loop 5 at temperature=1"
## [1] "Starting loop 6 at temperature=0.67"
## [1] "Current log likelihood=-54056.6006110304"
## [1] "Current log likelihood=-54056.6006110304"
## [1] "Completed loop 6 at temperature=0.67"
## [1] "Starting loop 7 at temperature=0"
## [1] "Current log likelihood=-35981.3444384101"
## [1] "Current log likelihood=-35981.3444384101"
## [1] "Completed loop 7 at temperature=0"</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co">#Add our results back into the Seurat object</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2">drag.clust &lt;-<span class="st"> </span>overall.clustering.soln[[<span class="st">"clusters"</span>]]</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">data.to.add &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/vector">vector</a></span>(<span class="dt">length=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(ser<span class="op">@</span>cell.names))</a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(data.to.add) &lt;-<span class="st"> </span>ser<span class="op">@</span>cell.names</a>
<a class="sourceLine" id="cb10-5" data-line-number="5">data.to.add &lt;-<span class="st"> </span>overall.clustering.soln[[<span class="st">"clusters"</span>]]</a>
<a class="sourceLine" id="cb10-6" data-line-number="6">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/AddMetaData">AddMetaData</a></span>(ser,<span class="dt">metadata=</span>data.to.add,<span class="dt">col.name=</span><span class="st">"DRAGON"</span>)</a>
<a class="sourceLine" id="cb10-7" data-line-number="7"></a>
<a class="sourceLine" id="cb10-8" data-line-number="8"><span class="co">#Visualize the final result vs the ground truth</span></a>
<a class="sourceLine" id="cb10-9" data-line-number="9">p1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/TSNEPlot">TSNEPlot</a></span>(ser,<span class="dt">group.by=</span><span class="st">"real_id"</span>,<span class="dt">do.return=</span>T) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">"Ground truth"</span>)</a>
<a class="sourceLine" id="cb10-10" data-line-number="10">p2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/TSNEPlot">TSNEPlot</a></span>(ser,<span class="dt">group.by=</span><span class="st">"DRAGON"</span>,<span class="dt">do.return=</span>T) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">"DRAGON"</span>)</a>
<a class="sourceLine" id="cb10-11" data-line-number="11"><span class="kw">plot_grid</span>(p1,p2)</a></code></pre></div>
<p><img src="dragonsc_files/figure-html/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Congratulations! You have performed your first clustering with DRAGON. Note that the clusters don’t align by name. Let’s fix that with BuildClusterTree in Seurat.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/SetAllIdent">SetAllIdent</a></span>(ser,<span class="dt">id=</span><span class="st">"real_id"</span>)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/BuildClusterTree">BuildClusterTree</a></span>(ser,<span class="dt">do.reorder=</span>T,<span class="dt">reorder.numeric=</span>T,<span class="dt">do.plot=</span><span class="ot">FALSE</span>,<span class="dt">show.progress=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb11-3" data-line-number="3">ser<span class="op">@</span>meta.data<span class="op">$</span>real_id &lt;-<span class="st"> </span>ser<span class="op">@</span>meta.data<span class="op">$</span>tree.ident</a>
<a class="sourceLine" id="cb11-4" data-line-number="4"></a>
<a class="sourceLine" id="cb11-5" data-line-number="5">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/SetAllIdent">SetAllIdent</a></span>(ser,<span class="dt">id=</span><span class="st">"DRAGON"</span>)</a>
<a class="sourceLine" id="cb11-6" data-line-number="6">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/BuildClusterTree">BuildClusterTree</a></span>(ser,<span class="dt">do.reorder=</span>T,<span class="dt">reorder.numeric=</span>T,<span class="dt">do.plot=</span><span class="ot">FALSE</span>,<span class="dt">show.progress=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb11-7" data-line-number="7">ser<span class="op">@</span>meta.data<span class="op">$</span>DRAGON &lt;-<span class="st"> </span>ser<span class="op">@</span>meta.data<span class="op">$</span>tree.ident</a>
<a class="sourceLine" id="cb11-8" data-line-number="8"></a>
<a class="sourceLine" id="cb11-9" data-line-number="9">p1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/TSNEPlot">TSNEPlot</a></span>(ser,<span class="dt">group.by=</span><span class="st">"real_id"</span>,<span class="dt">do.return=</span>T) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">"Ground truth"</span>)</a>
<a class="sourceLine" id="cb11-10" data-line-number="10">p2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/TSNEPlot">TSNEPlot</a></span>(ser,<span class="dt">group.by=</span><span class="st">"DRAGON"</span>,<span class="dt">do.return=</span>T) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">"DRAGON"</span>)</a>
<a class="sourceLine" id="cb11-11" data-line-number="11"><span class="kw">plot_grid</span>(p1,p2)</a></code></pre></div>
<p><img src="dragonsc_files/figure-html/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div id="compare-results-using-confusion-matrices-from-caret" class="section level1">
<h1 class="hasAnchor">
<a href="#compare-results-using-confusion-matrices-from-caret" class="anchor"></a>Compare results using confusion matrices from caret</h1>
<p>We can quantify the similarity between Ground truth and DRAGON using a confusion matrix from the caret package.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressMessages</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(caret))</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">caret<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/caret/topics/confusionMatrix">confusionMatrix</a></span>(<span class="dt">data=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">as.factor</a></span>(ser<span class="op">@</span>meta.data<span class="op">$</span>DRAGON),<span class="dt">reference=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">as.factor</a></span>(ser<span class="op">@</span>meta.data<span class="op">$</span>real_id))</a></code></pre></div>
<pre><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction    1    2    3    4    5    6
##          1  485    0    0    0    0    0
##          2    0 1251    0    0    0    0
##          3    0    0  750    0    0    0
##          4    0    0    0  495    0    0
##          5    0    0    0    0  761    0
##          6    0    0    0    0    0 1258
## 
## Overall Statistics
##                                      
##                Accuracy : 1          
##                  95% CI : (0.9993, 1)
##     No Information Rate : 0.2516     
##     P-Value [Acc &gt; NIR] : &lt; 2.2e-16  
##                                      
##                   Kappa : 1          
##                                      
##  Mcnemar's Test P-Value : NA         
## 
## Statistics by Class:
## 
##                      Class: 1 Class: 2 Class: 3 Class: 4 Class: 5 Class: 6
## Sensitivity             1.000   1.0000     1.00    1.000   1.0000   1.0000
## Specificity             1.000   1.0000     1.00    1.000   1.0000   1.0000
## Pos Pred Value          1.000   1.0000     1.00    1.000   1.0000   1.0000
## Neg Pred Value          1.000   1.0000     1.00    1.000   1.0000   1.0000
## Prevalence              0.097   0.2502     0.15    0.099   0.1522   0.2516
## Detection Rate          0.097   0.2502     0.15    0.099   0.1522   0.2516
## Detection Prevalence    0.097   0.2502     0.15    0.099   0.1522   0.2516
## Balanced Accuracy       1.000   1.0000     1.00    1.000   1.0000   1.0000</code></pre>
</div>
<div id="view-progressive-clustering-results-as-the-temperature-is-decreased" class="section level1">
<h1 class="hasAnchor">
<a href="#view-progressive-clustering-results-as-the-temperature-is-decreased" class="anchor"></a>View progressive clustering results as the temperature is decreased</h1>
<p>If you’ve chosen to save the intermediate results (as we have here by setting save.intermediate.results=TRUE), you can check out how clusters form as the temperature of the system is reduced. Annealing is cool! (Pun intended)</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"/Users/arc85/Desktop/dragon_intermediate_files"</span>)</a>
<a class="sourceLine" id="cb14-2" data-line-number="2"></a>
<a class="sourceLine" id="cb14-3" data-line-number="3">temp<span class="fl">.60</span> &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/readRDS">readRDS</a></span>(<span class="st">"em_results_temp_60.RDS"</span>)</a>
<a class="sourceLine" id="cb14-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(temp<span class="fl">.60</span>)</a></code></pre></div>
<pre><code>## [1] "pca.components" "expectation"    "maximization"   "clusters"      
## [5] "MLE"</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="co">#Clustering results are present in the "clusters" item in the named list</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2"></a>
<a class="sourceLine" id="cb16-3" data-line-number="3">drag.clust &lt;-<span class="st"> </span>temp<span class="fl">.60</span>[[<span class="st">"clusters"</span>]]</a>
<a class="sourceLine" id="cb16-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(drag.clust) &lt;-<span class="st"> </span>ser<span class="op">@</span>cell.names</a>
<a class="sourceLine" id="cb16-5" data-line-number="5">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/AddMetaData">AddMetaData</a></span>(ser,<span class="dt">metadata=</span>drag.clust,<span class="dt">col.name=</span><span class="st">"temp_60"</span>)</a>
<a class="sourceLine" id="cb16-6" data-line-number="6"></a>
<a class="sourceLine" id="cb16-7" data-line-number="7">temp<span class="fl">.4</span> &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/readRDS">readRDS</a></span>(<span class="st">"em_results_temp_4.RDS"</span>)</a>
<a class="sourceLine" id="cb16-8" data-line-number="8">drag.clust &lt;-<span class="st"> </span>temp<span class="fl">.4</span>[[<span class="st">"clusters"</span>]]</a>
<a class="sourceLine" id="cb16-9" data-line-number="9"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(drag.clust) &lt;-<span class="st"> </span>ser<span class="op">@</span>cell.names</a>
<a class="sourceLine" id="cb16-10" data-line-number="10">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/AddMetaData">AddMetaData</a></span>(ser,<span class="dt">metadata=</span>drag.clust,<span class="dt">col.name=</span><span class="st">"temp_4"</span>)</a>
<a class="sourceLine" id="cb16-11" data-line-number="11"></a>
<a class="sourceLine" id="cb16-12" data-line-number="12">temp<span class="fl">.1</span> &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/readRDS">readRDS</a></span>(<span class="st">"em_results_temp_1.RDS"</span>)</a>
<a class="sourceLine" id="cb16-13" data-line-number="13">drag.clust &lt;-<span class="st"> </span>temp<span class="fl">.1</span>[[<span class="st">"clusters"</span>]]</a>
<a class="sourceLine" id="cb16-14" data-line-number="14"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(drag.clust) &lt;-<span class="st"> </span>ser<span class="op">@</span>cell.names</a>
<a class="sourceLine" id="cb16-15" data-line-number="15">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/AddMetaData">AddMetaData</a></span>(ser,<span class="dt">metadata=</span>drag.clust,<span class="dt">col.name=</span><span class="st">"temp_1"</span>)</a>
<a class="sourceLine" id="cb16-16" data-line-number="16"></a>
<a class="sourceLine" id="cb16-17" data-line-number="17">p3 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/TSNEPlot">TSNEPlot</a></span>(ser,<span class="dt">group.by=</span><span class="st">"temp_60"</span>,<span class="dt">do.return=</span>T) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">"Temperature 60"</span>)</a>
<a class="sourceLine" id="cb16-18" data-line-number="18">p4 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/TSNEPlot">TSNEPlot</a></span>(ser,<span class="dt">group.by=</span><span class="st">"temp_4"</span>,<span class="dt">do.return=</span>T) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">"Temperature 4"</span>)</a>
<a class="sourceLine" id="cb16-19" data-line-number="19">p5 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/TSNEPlot">TSNEPlot</a></span>(ser,<span class="dt">group.by=</span><span class="st">"temp_1"</span>,<span class="dt">do.return=</span>T) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">"Temperature 1"</span>)</a>
<a class="sourceLine" id="cb16-20" data-line-number="20">p6 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/TSNEPlot">TSNEPlot</a></span>(ser,<span class="dt">group.by=</span><span class="st">"DRAGON"</span>,<span class="dt">do.return=</span>T) <span class="op">+</span></a>
<a class="sourceLine" id="cb16-21" data-line-number="21"><span class="kw">ggtitle</span>(<span class="st">"Temperature 0"</span>)</a>
<a class="sourceLine" id="cb16-22" data-line-number="22"></a>
<a class="sourceLine" id="cb16-23" data-line-number="23"><span class="kw">plot_grid</span>(p3,p4,p5,p6,<span class="dt">ncol=</span><span class="dv">2</span>)</a></code></pre></div>
<p><img src="dragonsc_files/figure-html/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#dragonsc-for-single-cell-rnaseq-clustering">DRAGONsc for single-cell RNAseq clustering</a></li>
      <li><a href="#simulate-data-using-splatter">Simulate data using splatter</a></li>
      <li><a href="#seurat-workflow-on-simulated-data">Seurat workflow on simulated data</a></li>
      <li><a href="#dragon-clustering">DRAGON clustering</a></li>
      <li><a href="#compare-results-using-confusion-matrices-from-caret">Compare results using confusion matrices from caret</a></li>
      <li><a href="#view-progressive-clustering-results-as-the-temperature-is-decreased">View progressive clustering results as the temperature is decreased</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Anthony Cillo.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
